How to deal with window-load timing in bootstrapped extensions

jikamens opened the pull request.

kewisch commented on add-ons/tb68/bootstrapped.md:
````diff
@@ -60,3 +60,65 @@ The shown example also specifies an optional `options` key to define the options
+```javascript
+const {Services} = ChromeUtils.import("resource://gre/modules/Services.jsm");
+
+var allInOneObserver = {
````
Can we make the examples here adhere to mail's eslint config?

jikamens replied:
I will try, but it would be nice if people who wanted to contribute docs to this repository did not have to clone the 320MB Thunderbird source tree and install the umpteen gazillion node modules needed to use eslint and the config in comm/.eslintrc.js.

kewisch commented on add-ons/tb68/bootstrapped.md:
```diff
@@ -60,3 +60,65 @@ The shown example also specifies an optional `options` key to define the options
+    // Alternatively, it might be easier to just check above in
+    // `handleDocument` to see if the document has the expected elements, and
+    // if not, then assume it's the wrong type of window and don't do anything.
+    var windows = Services.wm.getEnumerator(null);
```
Newer Gecko can do `for (let window of Services.wm.getEnumerator(null)) { ... }`, can you check if this is supported in 68 and adapt?

jikamens replied:
Done.

kewisch commented on add-ons/tb68/bootstrapped.md:
```diff
@@ -60,3 +60,65 @@ The shown example also specifies an optional `options` key to define the options
+
+function shutdown() {
+    Cc["@mozilla.org/embedcomp/window-watcher;1"].
+        getService(Ci.nsIWindowWatcher).unregisterNotification(
```
`Services.ww`

jikamens replied:
Done.

kewisch commented on add-ons/tb68/bootstrapped.md:
```diff
@@ -60,3 +60,65 @@ The shown example also specifies an optional `options` key to define the options
+    while (windows.hasMoreElements()) {
+        allInOneObserver.handleExistingWindow(windows.getNext());
+    }
+    Cc["@mozilla.org/embedcomp/window-watcher;1"].
```
`Services.ww`

jikamens replied:
Done.

jobisoft commented:
I would like to add three things:
1. I know for sure, that startup did not guarantee to wait until the final load of TB in TB60 as well, that is why I have always waited for the mail-startup-done observer.

2. The given example looks overly complicated/complex and authors probably cannot merge it into their bootstrap.js without much rework, specially if they use the window listener in a different way.

3. If the latest changes on TB core did make the underlying issue more observable and thus breaking addons which where lucky enough to have been working before, we should add some info.

In the mail in maildev you said, the mail-startup-done observer fixed it for you as well, which is what I would have expected, because you simple postpone all your registering, listening and what not until TB is ready. During install of the add-on you of course do not have to wait for that observer and just execute your stuff right away. That you have to inject your stuff into already open windows is not new and should be part of your standard routine, right?.

So I would have expected, that waiting for the mail-startup-done observer during startup is the only thing you need to add.

```
let onLoadObserver = {
  observe: function(aSubject, aTopic, aData) {
    // do your actual startup stuff
    ...
  }
}
function startup(data, reason) {
  Services.obs.addObserver(onLoadObserver, "mail-startup-done", false);
  if (reason != APP_STARTUP) {
    //during startup, we wait until mail-startup-done fired,
    //for all other reasons we call the observer directly
    onLoadObserver.observe();
  }
  // do not add anything her, but in onLoadObserver
}
```

jikamens commented:
>I know for sure, that startup did not guarantee to wait until the final load of TB in TB60 as well, that is why I have always waited for the mail-startup-done observer.

All I can tell you is that I never had this problem before and I maintain five different bootstrapped extensions. So whether the behavior was a fluke before or not, it indeed appears that "the latest changes on TB core did make the underlying issue more observable" and therefore "we should add some info," which is what I am proposing to do in this PR.

>In the mail in maildev you said, the mail-startup-done observer fixed it for you as well

`mail-startup-done` is sufficient if the only window you care about is the `mail:3pane` window that generates the `mail-startup-done` event. Some extensions -- including one of mine -- operate on other windows and therefore need to use the generic `load` event and check `document.readyState` as Geoff suggested on the list.

Also, your example code makes the assumption that during app startup the extension's `startup()` method will be called before the `mail-startup-done` event is generated. I'm not sure that assumption is any more reasonable than the opposite assumption, so the code has to be resilient to either possibility. My sample code is.

jikamens commented:
>The given example looks overly complicated/complex

If you have a less complicated/complex example that addresses all of the concerns I've outlined (handle all window types, handle `startup()` being called either before or after the window is fully loaded), I'm happy to use it. I don't see how to satisfy all of those requirements with simpler code.

If you don't actually think the example needs to satisfy those requirements, then that's certainly a conversation we can have, but I would tend to disagree.

jobisoft commented:
Instead of checking for `reason != APP_STARTUP`, one could check if the mail:3pane window is there. I will actually change that in my code.

Could you provide a link to the not-working-version of your add-on, so i could see what window you are talking about? However, I do not want to add more noise to this thread, maybe just send it to me in private.

jikamens commented:
>Instead of checking for `reason != APP_STARTUP`, one could check if the mail:3pane window is there. I will actually change that in my code.

Perhaps I am missing something, but I do not understand how this suggestion addresses either of the concerns I identified in my last comment.

>Could you provide a link to the not-working-version of your add-on, so i could see what window you are talking about?

It's not a specific window. It's _all_ windows, since userChromeJS loads the user's userChrome.js file for all windows. See https://github.com/jikamens/userChromeJS.

jobisoft commented:
I commented on this PR with the wrong mindset, assuming the reported issue is a timing issue of the startup() function inside bootstrap.js being called to early now. I somehow read that out of the maildev discussion.

Ignore what I posted.

jobisoft commented:
From an add-on author's point of view, it would be a good idea to have brief general information on this topic in the guide. As more and more authors work on their add-ons, I added a short preliminary section until someone agrees to this pull request.

jikamens commented:
I read what you added, and it is in my opinion insufficient.

I don't understand who the "someone" is whom I need to convince that this documentation should be added.

@kewisch said above that he isn't the right person, and @jobisoft seems to imply about that he isn't either, unless he's in the habit of speaking about himself vaguely in the third person?

Honestly I'm puzzled here. The problem I documented here is one that I actually encountered in the real world with two different bootstrapped extensions. Whether or not that's how it "should" be, that's how it is, and the odds are that if I encountered the problem, others are going to as well. What justification is there for not including what I proposed?

kewisch commented:
I think this is mainly up to the two of you to hash out and agree on. If you can't resolve it then I am sure @ryanleesipes can be the tiebreaker.

I am in favor of keeping the actual info short and then linking out to other guides. While you and others may have encountered this issue, most of the bootstrapped add-on examples I am aware of use the window loading code because it just isnt clear if the window is loaded. If this worked for your users you were likely just lucky the window loaded in time.

jikamens commented:
/me gives up in frustration
